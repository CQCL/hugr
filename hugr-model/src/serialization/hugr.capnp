@0xb4b3505e009086af;


using ParamId = UInt32;
using PortId = UInt32;
using TermId = UInt32;
using NodeId = UInt32;
using TermPtr = UInt32;

struct Node {
  # A node in the hugr graph.

  operation @0 :Operation;
  # The operation of the node.

  params @1 :List(ParamId);
  # Parameters that are passed to the operation.
  # Each parameter is an index to a term in the term table.

  inputs @2 :List(PortId);
  # Indices of the input ports of the node.

  outputs @3 :List(PortId);
  # Indices of the output ports of the node.

  children @4 :List(NodeId);
  # Children of the node.
  # Each child is an index to a node in the node table.

  meta @5 :List(MetaItem);
  # Metadata attached to the node.
}

struct Operation {
  union {
    module @0 :Void;
    input @1 :Void;
    output @2  :Void;
    dfg @3 :Void;
    cfg @4 :Void;
    block @5 :Void;
    case @6 :Void;
    exit @7 :Void;
    defineFunc @8 :DefineFunc;
    declareFunc @9 :DeclareFunc;
    callFunc @10 :CallFunc;
    loadFunc @11 :LoadFunc;
    custom @12 :Custom;
    defineAlias @13 :DefineAlias;
    declareAlias @14 :DeclareAlias;
  }

  struct DefineFunc {
    name @0 :Text;
    type @1 :Scheme;
  }

  struct DeclareFunc {
    name @0 :Text;
    type @1 :Scheme;
  }

  struct CallFunc {
    name @0 :Text;
    args @1 :List(UInt32);
  }

  struct LoadFunc {
    name @0 :Text;
  }

  struct Custom {
    name @0 :Text;
  }

  struct DefineAlias {
    name @0 :Symbol;
    value @1 :UInt32;
  }

  struct DeclareAlias {
    name @0 :Text;
    type @1 :UInt32;
  }
}

struct MetaItem {
  name @0 :Text;
  value @1 :Term;
}

struct Port {
  type @0 :UInt32;
  meta @1 :List(MetaItem);
}

struct Term {
  union {
    type @0 :Void;
    wildcard @1 :Void;
    var @2 :Text;
    named @3 :Named;
    list @4 :TypeList;
    listType @5 :ListType;
    str @6 :Text;
    strType @7 :Void;
    nat @8 :UInt64;
    natType @9 :Void;
    extSet @10 :ExtSet;
    extSetType @11 :Void;
    tuple @12 :Tuple;
    productType @13 :ProductType;
    tagged @14 :Tagged;
    sumType @15 :SumType;
    funcType @16 :FuncType;
  }

  struct Named {
    name @0 :Text;
    args @1 :List(UInt32);
  }

  struct TypeList {
    items @0 :List(UInt32);
    tail @1 :UInt32;
  }

  struct ListType {
    itemType @0 :UInt32;
  }

  struct Tuple {
    items @0 :List(UInt32);
  }

  struct ProductType {
    types @0 :UInt32;
  }

  struct FuncType {
    inputs @0 :UInt32;
    outputs @1 :UInt32;
    extensions @2 :UInt32;
  }

  struct SumType {
    types @0 :UInt32;
  }

  struct Tagged {
    tag @0 :UInt8;
    term @1 :UInt32;
  }

  struct ExtSet {
    extensions @0 :List(Text);
    rest @1 :UInt32;
  }
}

struct Scheme {
  params @0 :List(SchemeParam);
  constraints @1 :List(Constraint);
  body @2 :UInt32;
}

struct SchemeParam {
  name @0 :TermVar;
  type @1 :UInt32;
}

struct Constraint {
  name @0 :Symbol;
  args @1 :List(UInt32);
}

struct Symbol {
  name @0 :Text;
}

struct TermVar {
  name @0 :Text;
}


struct Module {
  nodes @0 :List(Node);
  ports @1 :List(Port);
  termTable @2 :List(Term);
  terms @3: List(TermPtr);
}
