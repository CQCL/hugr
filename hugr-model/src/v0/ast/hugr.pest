WHITESPACE  = _{ " " | "\t" | "\r" | "\n" }
COMMENT     = _{ ";" ~ (!("\n") ~ ANY)* ~ "\n" }
identifier  = @{ (ASCII_ALPHA | "_" | "-") ~ (ASCII_ALPHANUMERIC | "_" | "-")* }
symbol_name = @{ identifier ~ ("." ~ identifier)* }
link_name   = @{ "%" ~ (ASCII_ALPHANUMERIC | "_" | "-")* }

literal = { literal_string | literal_float | literal_nat | literal_bytes }

literal_string         = ${ "\"" ~ (literal_string_raw | literal_string_escape | literal_string_unicode)* ~ "\"" }
literal_string_raw     = @{ (!("\\" | "\"") ~ ANY)+ }
literal_string_escape  = @{ "\\" ~ ("\"" | "\\" | "n" | "r" | "t") }
literal_string_unicode = @{ "\\u" ~ "{" ~ ASCII_HEX_DIGIT+ ~ "}" }

literal_nat   = @{ (ASCII_DIGIT)+ }
literal_bytes =  { "(" ~ "bytes" ~ base64_string ~ ")" }
literal_float = @{ ("+" | "-")? ~ ASCII_DIGIT+ ~ "." ~ ASCII_DIGIT+ ~ ("e" ~ ("+" | "-")? ~ ASCII_DIGIT+)? }

base64_string = { "\"" ~ (ASCII_ALPHANUMERIC | "+" | "/")* ~ "="* ~ "\"" }

package = { "(" ~ "hugr" ~ "0" ~ ")" ~ module* ~ EOI }
module  = { "(" ~ "mod" ~ ")" ~ meta* ~ node* }

meta = { "(" ~ "meta" ~ term ~ ")" }

port_list  =  { "[" ~ link_name* ~ "]" }
port_lists = _{ port_list ~ port_list }

node = {
    node_dfg
  | node_cfg
  | node_block
  | node_define_func
  | node_declare_func
  | node_define_alias
  | node_declare_alias
  | node_declare_ctr
  | node_declare_operation
  | node_tail_loop
  | node_cond
  | node_import
  | node_custom
}

node_dfg               = { "(" ~ "dfg" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_cfg               = { "(" ~ "cfg" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_block             = { "(" ~ "block" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_define_func       = { "(" ~ "define-func" ~ symbol ~ meta* ~ region* ~ ")" }
node_declare_func      = { "(" ~ "declare-func" ~ symbol ~ meta* ~ ")" }
node_define_alias      = { "(" ~ "define-alias" ~ symbol ~ term ~ meta* ~ ")" }
node_declare_alias     = { "(" ~ "declare-alias" ~ symbol ~ meta* ~ ")" }
node_declare_ctr       = { "(" ~ "declare-ctr" ~ symbol ~ meta* ~ ")" }
node_declare_operation = { "(" ~ "declare-operation" ~ symbol ~ meta* ~ ")" }
node_tail_loop         = { "(" ~ "tail-loop" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_cond              = { "(" ~ "cond" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_import            = { "(" ~ "import" ~ symbol_name ~ meta* ~ ")" }
node_custom            = { "(" ~ term ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }

symbol = { symbol_name ~ param* ~ where_clause* ~ term }

signature    = { "(" ~ "signature" ~ term ~ ")" }
param        = { "(" ~ "param" ~ term_var ~ term ~ ")" }
where_clause = { "(" ~ "where" ~ term ~ ")" }

region      = { "(" ~ region_kind ~ port_lists? ~ signature? ~ meta* ~ node* ~ ")" }
region_kind = { "dfg" | "cfg" | "mod" }

term = {
    term_wildcard
  | term_var
  | term_list
  | term_list_type
  | literal
  | term_tuple
  | term_const_func
  | term_apply
}

term_wildcard   = @{ "_" ~ !(ASCII_ALPHANUMERIC | "_" | "-") }
term_var        =  { "?" ~ identifier }
term_apply      =  { symbol_name | ("(" ~ symbol_name ~ term* ~ ")") }
term_list       =  { "[" ~ part* ~ "]" }
term_tuple      =  { "(" ~ "tuple" ~ part* ~ ")" }
term_list_type  =  { "(" ~ "list" ~ term ~ ")" }
term_const_func =  { "(" ~ "fn" ~ term ~ ")" }

part         = { spliced_term | term }
spliced_term = { term ~ "..." }
