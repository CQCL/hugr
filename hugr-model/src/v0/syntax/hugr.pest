WHITESPACE  = _{ " " | "\t" | "\r" | "\n" }
COMMENT     = _{ ";" ~ (!("\n") ~ ANY)* ~ "\n" }
identifier  = @{ (ASCII_ALPHA | "_" | "-") ~ (ASCII_ALPHANUMERIC | "_" | "-")* }
ext_name    = @{ identifier ~ ("." ~ identifier)* }
symbol_name = @{ identifier ~ ("." ~ identifier)* }
link_name   = @{ "%" ~ (ASCII_ALPHANUMERIC | "_" | "-")* }

string         =  { "\"" ~ (string_raw | string_escape | string_unicode)* ~ "\"" }
string_raw     = @{ (!("\\" | "\"") ~ ANY)+ }
string_escape  = @{ "\\" ~ ("\"" | "\\" | "n" | "r" | "t") }
string_unicode = @{ "\\u" ~ "{" ~ ASCII_HEX_DIGIT+ ~ "}" }

base64_string = { "\"" ~ (ASCII_ALPHANUMERIC | "+" | "/")* ~ "="* ~ "\"" }

module = { "(" ~ "hugr" ~ "0" ~ ")" ~ meta* ~ node* ~ EOI }

meta = { "(" ~ "meta" ~ term ~ ")" }

port       =  { link_name }
port_list  =  { "[" ~ port* ~ "]" }
port_lists = _{ port_list ~ port_list }

node = {
    node_dfg
  | node_cfg
  | node_block
  | node_define_func
  | node_declare_func
  | node_define_alias
  | node_declare_alias
  | node_declare_ctr
  | node_declare_operation
  | node_tail_loop
  | node_cond
  | node_import
  | node_custom
}

node_dfg               = { "(" ~ "dfg" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_cfg               = { "(" ~ "cfg" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_block             = { "(" ~ "block" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_define_func       = { "(" ~ "define-func" ~ symbol ~ meta* ~ region* ~ ")" }
node_declare_func      = { "(" ~ "declare-func" ~ symbol ~ meta* ~ ")" }
node_define_alias      = { "(" ~ "define-alias" ~ symbol ~ term ~ meta* ~ ")" }
node_declare_alias     = { "(" ~ "declare-alias" ~ symbol ~ meta* ~ ")" }
node_declare_ctr       = { "(" ~ "declare-ctr" ~ symbol ~ meta* ~ ")" }
node_declare_operation = { "(" ~ "declare-operation" ~ symbol ~ meta* ~ ")" }
node_tail_loop         = { "(" ~ "tail-loop" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_cond              = { "(" ~ "cond" ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }
node_import            = { "(" ~ "import" ~ symbol_name ~ meta* ~ ")" }
node_custom            = { "(" ~ term_apply ~ port_lists? ~ signature? ~ meta* ~ region* ~ ")" }

symbol = { symbol_name ~ param* ~ where_clause* ~ term }

signature    = { "(" ~ "signature" ~ term ~ ")" }
param        = { "(" ~ "param" ~ term_var ~ term ~ ")" }
where_clause = { "(" ~ "where" ~ term ~ ")" }

region      = { "(" ~ region_kind ~ port_lists? ~ signature? ~ meta* ~ node* ~ ")" }
region_kind = { "dfg" | "cfg" | "mod" }

term = {
    term_wildcard
  | term_var
  | term_list
  | term_list_type
  | term_str
  | term_float
  | term_nat
  | term_ext_set
  | term_apply
  | term_const_func
  | term_bytes
  | term_float
  | term_tuple
}

term_wildcard   =  { "_" }
term_var        =  { "?" ~ identifier }
term_apply      =  { symbol_name | ("(" ~ symbol_name ~ term* ~ ")") }
term_list       =  { "[" ~ part* ~ "]" }
term_tuple      =  { "(" ~ "tuple" ~ part* ~ ")" }
term_list_type  =  { "(" ~ "list" ~ term ~ ")" }
term_str        =  { string }
term_nat        = @{ (ASCII_DIGIT)+ }
term_ext_set    =  { "(" ~ "ext" ~ (spliced_term | ext_name)* ~ ")" }
term_const_func =  { "(" ~ "fn" ~ term ~ ")" }
term_bytes      =  { "(" ~ "bytes" ~ base64_string ~ ")" }
term_float      = @{ ("+" | "-")? ~ (ASCII_DIGIT)+ ~ "." ~ (ASCII_DIGIT)+ }

part         = { spliced_term | term }
spliced_term = { term ~ "..." }
