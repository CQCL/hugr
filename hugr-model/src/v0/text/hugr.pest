WHITESPACE = _{ " " | "\t" | "\r" | "\n" }
COMMENT    = _{ ";" ~ (!("\n") ~ ANY)* ~ "\n" }
identifier = @{ (ASCII_ALPHA | "_" | "-") ~ (ASCII_ALPHANUMERIC | "_" | "-")* }
ext_name   = @{ identifier ~ ("." ~ identifier)* }
symbol     = @{ identifier ~ ("." ~ identifier)+ }
tag        = @{ (ASCII_NONZERO_DIGIT ~ ASCII_DIGIT*) | "0" }

string    = @{ "\"" ~ (!("\"") ~ ANY)* ~ "\"" }
list_tail =  { "." }

module = { "(" ~ "hugr" ~ "0" ~ ")" ~ meta* ~ node* ~ EOI }

meta = { "(" ~ "meta" ~ symbol ~ term ~ ")" }

edge_name  = @{ "%" ~ (ASCII_ALPHANUMERIC | "_" | "-")* }
port       =  { edge_name | ("(" ~ edge_name ~ term ~ meta* ~ ")") }
port_list  =  { "[" ~ port* ~ "]" }
port_lists = _{ port_list ~ port_list }

node = {
    node_dfg
  | node_cfg
  | node_block
  | node_define_func
  | node_declare_func
  | node_call_func
  | node_load_func
  | node_define_alias
  | node_declare_alias
  | node_tail_loop
  | node_cond
  | node_tag
  | node_custom
}

node_dfg           = { "(" ~ "dfg" ~ port_lists? ~ type_hint? ~ meta* ~ region* ~ ")" }
node_cfg           = { "(" ~ "cfg" ~ port_lists? ~ type_hint? ~ meta* ~ region* ~ ")" }
node_block         = { "(" ~ "block" ~ port_lists? ~ type_hint? ~ meta* ~ region* ~ ")" }
node_define_func   = { "(" ~ "define-func" ~ func_header ~ meta* ~ region* ~ ")" }
node_declare_func  = { "(" ~ "declare-func" ~ func_header ~ meta* ~ ")" }
node_call_func     = { "(" ~ "call" ~ term ~ port_lists? ~ type_hint? ~ meta* ~ ")" }
node_load_func     = { "(" ~ "load-func" ~ term ~ port_lists? ~ type_hint? ~ meta* ~ ")" }
node_define_alias  = { "(" ~ "define-alias" ~ alias_header ~ term ~ meta* ~ ")" }
node_declare_alias = { "(" ~ "declare-alias" ~ alias_header ~ meta* ~ ")" }
node_tail_loop     = { "(" ~ "tail-loop" ~ port_lists? ~ type_hint? ~ meta* ~ region* ~ ")" }
node_cond          = { "(" ~ "cond" ~ port_lists? ~ type_hint? ~ meta* ~ region* ~ ")" }
node_tag           = { "(" ~ "tag" ~ tag ~ port_lists? ~ type_hint? ~ meta* ~ region* ~ ")" }
node_custom        = { "(" ~ (term_apply | term_apply_full) ~ port_lists? ~ type_hint? ~ meta* ~ region* ~ ")" }

type_hint    = { "(" ~ "type" ~ term ~ ")" }
func_header  = { symbol ~ param* ~ term ~ term ~ term }
alias_header = { symbol ~ param* ~ term }

param = { param_implicit | param_explicit | param_constraint }

param_implicit   = { "(" ~ "forall" ~ term_var ~ term ~ ")" }
param_explicit   = { "(" ~ "param" ~ term_var ~ term ~ ")" }
param_constraint = { "(" ~ "where" ~ term ~ ")" }

region     = { region_dfg | region_cfg }
region_dfg = { "(" ~ "dfg" ~ port_lists? ~ type_hint? ~ meta* ~ node* ~ ")" }
region_cfg = { "(" ~ "cfg" ~ port_lists? ~ type_hint? ~ meta* ~ node* ~ ")" }

term = {
    term_wildcard
  | term_type
  | term_static
  | term_constraint
  | term_var
  | term_quote
  | term_list
  | term_list_type
  | term_str
  | term_str_type
  | term_nat
  | term_nat_type
  | term_ext_set
  | term_ext_set_type
  | term_adt
  | term_func_type
  | term_ctrl
  | term_ctrl_type
  | term_apply_full
  | term_apply
}

term_wildcard     = { "_" }
term_type         = { "type" }
term_static       = { "static" }
term_constraint   = { "constraint" }
term_var          = { "?" ~ identifier }
term_apply_full   = { ("(" ~ "@" ~ symbol ~ term* ~ ")") }
term_apply        = { symbol | ("(" ~ symbol ~ term* ~ ")") }
term_quote        = { "(" ~ "quote" ~ term ~ ")" }
term_list         = { "[" ~ term* ~ (list_tail ~ term)? ~ "]" }
term_list_type    = { "(" ~ "list" ~ term ~ ")" }
term_str          = { string }
term_str_type     = { "str" }
term_nat          = { (ASCII_DIGIT)+ }
term_nat_type     = { "nat" }
term_ext_set      = { "(" ~ "ext" ~ ext_name* ~ (list_tail ~ term)? ~ ")" }
term_ext_set_type = { "ext-set" }
term_adt          = { "(" ~ "adt" ~ term ~ ")" }
term_func_type    = { "(" ~ "fn" ~ term ~ term ~ term ~ ")" }
term_ctrl         = { "(" ~ "ctrl" ~ term ~ ")" }
term_ctrl_type    = { "ctrl" }
