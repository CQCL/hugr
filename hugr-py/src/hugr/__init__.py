"""`hugr` is a Python package for the Quantinuum HUGR common
representation.
"""

__version__ = "0.1.0"


def it_works() -> str:
    """Return a string to confirm that the package is installed and working."""
    return "It works!"
