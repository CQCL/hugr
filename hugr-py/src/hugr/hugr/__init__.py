"""The main HUGR structure."""

from .base import Hugr, NodeData
from .node_port import (
    Direction,
    InPort,
    Node,
    OutPort,
    Wire,
)

__all__ = [
    "Hugr",
    "NodeData",
    "Direction",
    "InPort",
    "Wire",
    "OutPort",
    "Node",
]
