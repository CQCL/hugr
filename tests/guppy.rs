use std::{
    env,
    fmt::Display,
    fs::File,
    path::{Path, PathBuf},
};

use anyhow::Result;
use hugr::{
    extension::{prelude, ExtensionId, ExtensionRegistry},
    ops::{DataflowOpTrait as _, ExtensionOp},
    std_extensions::arithmetic::{float_types, int_ops, int_types},
    types::CustomType,
    Hugr, HugrView,
};
use hugr_llvm::{
    custom::{CodegenExtension, CodegenExtsMap},
    emit::{deaggregate_call_result, EmitFuncContext, EmitHugr, EmitOpArgs, Namer},
    fat::FatExt as _,
    types::TypingSession,
};
use inkwell::{context::Context, module::Module, types::BasicTypeEnum};
use itertools::Itertools as _;
use lazy_static::lazy_static;
use rstest::{fixture, rstest};

lazy_static! {
    static ref EXTENSION_REGISTRY: ExtensionRegistry = ExtensionRegistry::try_new([
        int_ops::EXTENSION.to_owned(),
        int_types::EXTENSION.to_owned(),
        prelude::PRELUDE.to_owned(),
        float_types::EXTENSION.to_owned(),
    ])
    .unwrap();
}

// A toy codegen extension for "quantum.tket2" ops.
struct Tket2CodegenExtension;

impl<H: HugrView> CodegenExtension<H> for Tket2CodegenExtension {
    fn extension(&self) -> ExtensionId {
        ExtensionId::new("quantum.tket2").unwrap()
    }

    fn llvm_type<'c>(
        &self,
        _context: &TypingSession<'c, H>,
        _hugr_type: &CustomType,
    ) -> anyhow::Result<BasicTypeEnum<'c>> {
        unimplemented!()
    }

    fn emit_extension_op<'c>(
        &self,
        context: &mut EmitFuncContext<'c, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
    ) -> Result<()> {
        // we lower all ops by declaring an extern function of the same name
        // and signature, and calling that function.
        // let opaque = args.node().as_ref().clone().into_opaque();
        let node = args.node();
        let sig = node.signature();
        let func_type = context.llvm_func_type(&sig)?;
        let func = context.get_extern_func(node.def().name(), func_type)?;
        let call_args = args.inputs.into_iter().map_into().collect_vec();
        let builder = context.builder();
        let call_result = builder.build_call(func, &call_args, "")?;
        let call_result = deaggregate_call_result(builder, call_result, args.outputs.len())?;
        args.outputs.finish(builder, call_result)
    }
}

// drives `hugr-llvm` to produce an LLVM module from a Hugr.
fn hugr_to_module<'c>(context: &'c Context, hugr: &'c Hugr) -> Module<'c> {
    let module = context.create_module("test_context");
    let namer = Namer::default().into();
    let exts = CodegenExtsMap::default()
        .add_int_extensions()
        .add_logic_extensions()
        .add_default_prelude_extensions()
        .add_float_extensions()
        .add_cge(Tket2CodegenExtension);
    let root = hugr.fat_root().unwrap();
    EmitHugr::new(context, module, namer, exts.into())
        .emit_module(root)
        .unwrap()
        .finish()
}

struct TestConfig {
    test_dir: PathBuf,
}

impl TestConfig {
    pub fn new() -> TestConfig {
        let test_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/guppy_test_cases");
        assert!(test_dir.is_dir());
        TestConfig { test_dir }
    }
}

impl TestConfig {
    // invokes the path with python, expecting to recieve Hugr serialised as
    // JSON on stdout.
    fn get_guppy_output(&self, path: impl AsRef<Path>) -> Hugr {
        let mut hugr: Hugr =
            serde_json::from_reader(File::open(self.test_dir.join(path)).unwrap()).unwrap();
        hugr.update_validate(&EXTENSION_REGISTRY).unwrap();
        hugr
    }

    fn run<T>(&self, path: impl AsRef<Path>, opt: bool, go: impl FnOnce(String) -> T) -> T {
        let hugr = self.get_guppy_output(path);
        let context = Context::create();
        let module = hugr_to_module(&context, &hugr);
        module
            .verify()
            .unwrap_or_else(|pp| panic!("Failed to verify module: {pp}"));
        if opt {
            let pb = inkwell::passes::PassManager::create(());
            pb.add_promote_memory_to_register_pass();
            pb.run_on(&module);
        }
        go(module.to_string())
    }
}

#[fixture]
fn test_config() -> TestConfig {
    TestConfig::new()
}

fn with_suffix<R>(s: impl Display, go: impl FnOnce() -> R) -> R {
    let mut settings = insta::Settings::clone_current();
    let old_suffix = settings
        .snapshot_suffix()
        .map_or("".to_string(), |s| format!("{s}."));
    let llvm_str = hugr_llvm::llvm_version();
    settings.set_snapshot_suffix(format!("{old_suffix}{llvm_str}.{s}"));
    settings.bind(go)
}

macro_rules! guppy_test {
    ($filename:expr, $testname:ident) => {
        #[rstest]
        #[ignore = "Guppy has not yet been upgraded to hugr-0.12.0"]
        fn $testname(test_config: TestConfig) {
            with_suffix("noopt", || {
                test_config.run($filename, false, |module_string| {
                    insta::assert_snapshot!(module_string)
                });
            });
            with_suffix("opt", || {
                test_config.run($filename, true, |module_string| {
                    insta::assert_snapshot!(module_string)
                });
            });
        }
    };
}

guppy_test!("even_odd.py.json", even_odd);
guppy_test!("even_odd2.py.json", even_odd2);
guppy_test!("planqc-1.py.json", planqc1);
guppy_test!("planqc-2.py.json", planqc2);
guppy_test!("planqc-3.py.json", planqc3);
