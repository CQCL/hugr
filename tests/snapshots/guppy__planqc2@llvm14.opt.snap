---
source: tests/guppy.rs
expression: module_string
---
; ModuleID = 'test_context'
source_filename = "test_context"

define i16 @_hl.teleport.25(i16 %0, i16 %1) {
alloca_block:
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  br label %2

2:                                                ; preds = %entry_block
  %3 = call i16 @QAlloc()
  %4 = call i16 @H(i16 %3)
  %5 = call { i16, i16 } @CX(i16 %4, i16 %1)
  %6 = extractvalue { i16, i16 } %5, 0
  %7 = extractvalue { i16, i16 } %5, 1
  %8 = insertvalue { i16, i16 } undef, i16 %6, 0
  %9 = insertvalue { i16, i16 } %8, i16 %7, 1
  %10 = insertvalue { { i16, i16 } } poison, { i16, i16 } %9, 0
  %11 = extractvalue { { i16, i16 } } %10, 0
  %12 = extractvalue { i16, i16 } %11, 0
  %13 = extractvalue { i16, i16 } %11, 1
  %14 = call { i16, i16 } @CX(i16 %0, i16 %12)
  %15 = extractvalue { i16, i16 } %14, 0
  %16 = extractvalue { i16, i16 } %14, 1
  %17 = insertvalue { i16, i16 } undef, i16 %15, 0
  %18 = insertvalue { i16, i16 } %17, i16 %16, 1
  %19 = insertvalue { { i16, i16 } } poison, { i16, i16 } %18, 0
  %20 = extractvalue { { i16, i16 } } %19, 0
  %21 = extractvalue { i16, i16 } %20, 0
  %22 = extractvalue { i16, i16 } %20, 1
  %23 = call i16 @H(i16 %21)
  %24 = call { i16, { i32, {}, {} } } @Measure(i16 %23)
  %25 = extractvalue { i16, { i32, {}, {} } } %24, 0
  %26 = extractvalue { i16, { i32, {}, {} } } %24, 1
  call void @QFree(i16 %25)
  %27 = extractvalue { i32, {}, {} } %26, 0
  switch i32 %27, label %28 [
    i32 1, label %30
  ]

28:                                               ; preds = %2
  %29 = extractvalue { i32, {}, {} } %26, 1
  br label %36

30:                                               ; preds = %2
  %31 = extractvalue { i32, {}, {} } %26, 2
  br label %32

32:                                               ; preds = %30
  %33 = call i16 @Z(i16 %13)
  switch i32 0, label %34 [
  ]

34:                                               ; preds = %32
  %35 = extractvalue { {} } undef, 0
  br label %39

36:                                               ; preds = %28
  switch i32 0, label %37 [
  ]

37:                                               ; preds = %36
  %38 = extractvalue { {} } undef, 0
  br label %39

39:                                               ; preds = %37, %34
  %"9_0.0" = phi i16 [ %13, %37 ], [ %33, %34 ]
  %40 = call { i16, { i32, {}, {} } } @Measure(i16 %22)
  %41 = extractvalue { i16, { i32, {}, {} } } %40, 0
  %42 = extractvalue { i16, { i32, {}, {} } } %40, 1
  call void @QFree(i16 %41)
  %43 = extractvalue { i32, {}, {} } %42, 0
  switch i32 %43, label %44 [
    i32 1, label %46
  ]

44:                                               ; preds = %39
  %45 = extractvalue { i32, {}, {} } %42, 1
  br label %52

46:                                               ; preds = %39
  %47 = extractvalue { i32, {}, {} } %42, 2
  br label %48

48:                                               ; preds = %46
  %49 = call i16 @X(i16 %"9_0.0")
  switch i32 0, label %50 [
  ]

50:                                               ; preds = %48
  %51 = extractvalue { {} } undef, 0
  br label %55

52:                                               ; preds = %44
  switch i32 0, label %53 [
  ]

53:                                               ; preds = %52
  %54 = extractvalue { {} } undef, 0
  br label %55

55:                                               ; preds = %53, %50
  %"12_0.0" = phi i16 [ %"9_0.0", %53 ], [ %49, %50 ]
  switch i32 0, label %56 [
  ]

56:                                               ; preds = %55
  %57 = extractvalue { {} } undef, 0
  br label %58

58:                                               ; preds = %56
  ret i16 %"12_0.0"
}

declare i16 @QAlloc()

declare i16 @H(i16)

declare { i16, i16 } @CX(i16, i16)

declare { i16, { i32, {}, {} } } @Measure(i16)

declare void @QFree(i16)

declare i16 @Z(i16)

declare i16 @X(i16)

define { i32, {}, {} } @_hl.main.26() {
alloca_block:
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  br label %0

0:                                                ; preds = %entry_block
  %1 = call i16 @QAlloc()
  %2 = call i16 @QAlloc()
  %3 = insertvalue { i16, i16 } undef, i16 %2, 0
  %4 = insertvalue { i16, i16 } %3, i16 %1, 1
  %5 = insertvalue { { i16, i16 } } poison, { i16, i16 } %4, 0
  %6 = extractvalue { { i16, i16 } } %5, 0
  %7 = extractvalue { i16, i16 } %6, 0
  %8 = extractvalue { i16, i16 } %6, 1
  %9 = call i16 @_hl.teleport.25(i16 %7, i16 %8)
  %10 = call { i16, { i32, {}, {} } } @Measure(i16 %9)
  %11 = extractvalue { i16, { i32, {}, {} } } %10, 0
  %12 = extractvalue { i16, { i32, {}, {} } } %10, 1
  call void @QFree(i16 %11)
  switch i32 0, label %13 [
  ]

13:                                               ; preds = %0
  %14 = extractvalue { {} } undef, 0
  br label %15

15:                                               ; preds = %13
  ret { i32, {}, {} } %12
}
